File: Outputs/Logistic Regression/Logistic Regression.rs
Training Logistic Regression...
[Parallel(n_jobs=-1)]: Using backend LokyBackend with 8 concurrent workers.
Training completed in 31.14 seconds.
Running predictions...
Inference completed in 0.03 seconds.
Logistic Regression Accuracy: 92.21%

Average Number of Activated Pixels for Each Digit:
Digit 0: 191.97 pixels
Digit 1: 85.85 pixels
Digit 2: 168.81 pixels
Digit 3: 163.34 pixels
Digit 4: 141.81 pixels
Digit 5: 152.33 pixels
Digit 6: 156.93 pixels
Digit 7: 131.40 pixels
Digit 8: 173.32 pixels
Digit 9: 143.03 pixels

2025-02-03 15:19:42.580 Python[74546:9498147] +[IMKClient subclass]: chose IMKClient_Modern
2025-02-03 15:19:42.580 Python[74546:9498147] +[IMKInputSession subclass]: chose IMKInputSession_Modern
Classification Report:
              precision    recall  f1-score   support

           0       0.95      0.97      0.96       980
           1       0.96      0.98      0.97      1135
           2       0.92      0.89      0.90      1032
           3       0.90      0.91      0.91      1010
           4       0.94      0.93      0.93       982
           5       0.89      0.87      0.88       892
           6       0.94      0.95      0.95       958
           7       0.93      0.92      0.92      1028
           8       0.87      0.88      0.88       974
           9       0.91      0.92      0.91      1009

    accuracy                           0.92     10000
   macro avg       0.92      0.92      0.92     10000
weighted avg       0.92      0.92      0.92     10000

Logistic Regression Model Training Time: 31.14 seconds
Logistic Regression Model Testing Time: 0.03 seconds
Logistic Regression Model Test Accuracy: 92.21%

Misclassified counts by digit for Logistic Regression:
Digit 0: 32
Digit 1: 27
Digit 2: 112
Digit 3: 88
Digit 4: 67
Digit 5: 118
Digit 6: 48
Digit 7: 84
Digit 8: 118
Digit 9: 85

Successful classification percentage for Logistic Regression by digit:
Digit 0: 96.73%
Digit 1: 97.62%
Digit 2: 89.15%
Digit 3: 91.29%
Digit 4: 93.18%
Digit 5: 86.77%
Digit 6: 94.99%
Digit 7: 91.83%
Digit 8: 87.89%
Digit 9: 91.58%

File: Models/Logistic Regression.py
import numpy as np
import matplotlib.pyplot as plt
from tensorflow.keras.datasets import mnist # type: ignore
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix, classification_report, accuracy_score
from sklearn.preprocessing import StandardScaler
import seaborn as sns
import time

# Load MNIST dataset
(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

# Flatten and normalize pixel values
train_images_flat = train_images.reshape(train_images.shape[0], -1)
test_images_flat = test_images.reshape(test_images.shape[0], -1)

scaler = StandardScaler()
train_images_flat = scaler.fit_transform(train_images_flat)
test_images_flat = scaler.transform(test_images_flat)

# Compute the average number of activated pixels per digit (nonzero pixels)
average_active_pixels = {
    digit: np.mean([np.count_nonzero(img) for img in train_images[train_labels == digit]])
    for digit in range(10)
}

# Initialize Logistic Regression with tuned settings
model_lr = LogisticRegression(
    max_iter=500,          # Lower max iterations to prevent excessive time
    solver='lbfgs',        # Optimized solver for multi-class classification
    n_jobs=-1,             # Use all CPU cores
    verbose=1              # Enable verbose output to track training progress
)

# Measure training time
print("Training Logistic Regression...")
start_time = time.time()
model_lr.fit(train_images_flat, train_labels)
training_time = time.time() - start_time
print(f"Training completed in {training_time:.2f} seconds.")

# Measure inference time
print("Running predictions...")
start_time = time.time()
lr_predictions = model_lr.predict(test_images_flat)
inference_time = time.time() - start_time
print(f"Inference completed in {inference_time:.2f} seconds.")

# Evaluate accuracy
lr_accuracy = accuracy_score(test_labels, lr_predictions)
print(f"Logistic Regression Accuracy: {lr_accuracy * 100:.2f}%")

# Print the average number of activated pixels for each digit
print("\nAverage Number of Activated Pixels for Each Digit:")
for digit, avg_pixels in average_active_pixels.items():
    print(f"Digit {digit}: {avg_pixels:.2f} pixels")

# Confusion Matrix (to identify frequently misclassified digits)
cm = confusion_matrix(test_labels, lr_predictions)
plt.figure(figsize=(10, 8))
sns.heatmap(cm, annot=True, fmt='d', cmap='Blues', cbar=False)
plt.xlabel('Predicted')
plt.ylabel('True')
plt.title('Confusion Matrix for Logistic Regression')
plt.show()

# Classification Report (precision, recall, F1-score)
print("Classification Report:")
print(classification_report(test_labels, lr_predictions))

# Visualize misclassified digits for each class
misclassified = np.where(test_labels != lr_predictions)[0]
misclassified_by_class = {i: [] for i in range(10)}

# Group misclassified samples by true label
for idx in misclassified:
    true_label = test_labels[idx]
    misclassified_by_class[true_label].append(idx)

# Plot one misclassified example per class
plt.figure(figsize=(15, 8))
for i in range(10):
    if misclassified_by_class[i]:  
        idx = misclassified_by_class[i][0]  
        plt.subplot(2, 5, i+1)
        plt.imshow(test_images[idx].reshape(28, 28), cmap='gray')
        plt.title(f"True: {test_labels[idx]}\nPred: {lr_predictions[idx]}")
    else:
        plt.subplot(2, 5, i+1)
        plt.text(0.5, 0.5, f"No misclassified\nfor {i}", ha='center', va='center')
        plt.axis('off')
plt.suptitle('Misclassified Digits by Logistic Regression (One Example per Class)')
plt.show()

# Calculate misclassified counts by digit
misclassified_counts = {i: 0 for i in range(10)}
for idx in misclassified:
    true_label = test_labels[idx]
    misclassified_counts[true_label] += 1

# Calculate successful classification percentage by digit
success_percentage = {i: 0 for i in range(10)}
for i in range(10):
    total_samples = np.sum(test_labels == i)
    correct_samples = total_samples - misclassified_counts[i]
    success_percentage[i] = (correct_samples / total_samples) * 100

# Print results
print(f"Logistic Regression Model Training Time: {training_time:.2f} seconds")
print(f"Logistic Regression Model Testing Time: {inference_time:.2f} seconds")
print(f"Logistic Regression Model Test Accuracy: {lr_accuracy * 100:.2f}%")
print("\nMisclassified counts by digit for Logistic Regression:")
for i in range(10):
    print(f"Digit {i}: {misclassified_counts[i]}")
print("\nSuccessful classification percentage for Logistic Regression by digit:")
for i in range(10):
    print(f"Digit {i}: {success_percentage[i]:.2f}%")
